### PR TITLE
[IR][Backend] Generate and wrap the top-level data movement

### PR DESCRIPTION
**Pull request by AdrianLiu00**
## Description ##
This PR enhances the functionality of generating top-level buffers and managing relevant data movement for the _vitis_hls_ target. Additionally, it optimizes some code logic within the generation process.

### Problems ###
The original top-level load and store logic results in incorrect segment behavior for the _#dataflow_ primitive.

### Proposed Solutions ###
This PR addresses the issue by explicitly generating and encapsulating data movement into separate functions, which are then invoked within the top-level function. As a result, the top-level function is simplified to consist solely of function calls and variable definitions, ensuring correct behavior under the _#dataflow_ primitive.

### Examples ###
Original:
```
  func.func @top(%arg0: memref<1024xf32>, %arg1: memref<1024xf32>, %arg2: memref<1024xf32>) attributes {dataflow} {
    %alloc = memref.alloc() {name = "buf0"} : memref<32x32xf32>
    affine.for %arg3 = 0 to 32 {
      affine.for %arg4 = 0 to 32 {
        %0 = affine.load %arg0[%arg3 * 32 + %arg4] : memref<1024xf32>
        affine.store %0, %alloc[%arg3, %arg4] : memref<32x32xf32>
      } {loop_name = "buf0_l_1", pipeline_ii = 1 : ui32, rewind}
    } {loop_name = "buf0_l_0", op_name = "S_buf0"}
    %alloc_0 = memref.alloc() {name = "buf1"} : memref<32x32xf32>
    affine.for %arg3 = 0 to 32 {
      affine.for %arg4 = 0 to 32 {
        %0 = affine.load %arg1[%arg3 * 32 + %arg4] : memref<1024xf32>
        affine.store %0, %alloc_0[%arg3, %arg4] : memref<32x32xf32>
      } {loop_name = "buf1_l_1", pipeline_ii = 1 : ui32, rewind}
    } {loop_name = "buf1_l_0", op_name = "S_buf1"}
    %alloc_1 = memref.alloc() {name = "buf2"} : memref<32x32xf32>
    affine.for %arg3 = 0 to 32 {
      affine.for %arg4 = 0 to 32 {
        %0 = affine.load %arg2[%arg3 * 32 + %arg4] : memref<1024xf32>
        affine.store %0, %alloc_1[%arg3, %arg4] : memref<32x32xf32>
      } {loop_name = "buf2_l_1", pipeline_ii = 1 : ui32, rewind}
    } {loop_name = "buf2_l_0", op_name = "S_buf2"}
    call @gemm_0_0(%alloc, %alloc_0, %alloc_1) : (memref<32x32xf32>, memref<32x32xf32>, memref<32x32xf32>) -> ()
    call @gemm_0_1(%alloc, %alloc_0, %alloc_1) : (memref<32x32xf32>, memref<32x32xf32>, memref<32x32xf32>) -> ()
    call @gemm_1_0(%alloc, %alloc_0, %alloc_1) : (memref<32x32xf32>, memref<32x32xf32>, memref<32x32xf32>) -> ()
    call @gemm_1_1(%alloc, %alloc_0, %alloc_1) {last} : (memref<32x32xf32>, memref<32x32xf32>, memref<32x32xf32>) -> ()
    affine.for %arg3 = 0 to 32 {
      affine.for %arg4 = 0 to 32 {
        %0 = affine.load %alloc_1[%arg3, %arg4] : memref<32x32xf32>
        affine.store %0, %arg2[%arg3 * 32 + %arg4] : memref<1024xf32>
      } {loop_name = "result3_l_1", pipeline_ii = 1 : ui32, rewind}
    } {loop_name = "result3_l_0", op_name = "S_result3"}
    return
  }
```

Now:
```
  ...
  func.func @load_buf2(%arg0: memref<1024xf32>, %arg1: memref<32x32xf32>) {
    affine.for %arg2 = 0 to 32 {
      affine.for %arg3 = 0 to 32 {
        %0 = affine.load %arg0[%arg2 * 32 + %arg3] : memref<1024xf32>
        affine.store %0, %arg1[%arg2, %arg3] : memref<32x32xf32>
      } {loop_name = "load_buf2_l_1", pipeline_ii = 1 : ui32, rewind}
    } {loop_name = "load_buf2_l_0", op_name = "S_load_buf2"}
    return
  }
  func.func @store_res(%arg0: memref<32x32xf32>, %arg1: memref<1024xf32>) {
    affine.for %arg2 = 0 to 32 {
      affine.for %arg3 = 0 to 32 {
        %0 = affine.load %arg0[%arg2, %arg3] : memref<32x32xf32>
        affine.store %0, %arg1[%arg2 * 32 + %arg3] : memref<1024xf32>
      } {loop_name = "store_res_l_1", pipeline_ii = 1 : ui32, rewind}
    } {loop_name = "store_res_l_0", op_name = "S_store_res"}
    return
  }
  func.func @top(%arg0: memref<1024xf32>, %arg1: memref<1024xf32>, %arg2: memref<1024xf32>) attributes {dataflow} {
    %alloc = memref.alloc() : memref<32x32xf32>
    call @load_buf0(%arg0, %alloc) : (memref<1024xf32>, memref<32x32xf32>) -> ()
    %alloc_0 = memref.alloc() : memref<32x32xf32>
    call @load_buf1(%arg1, %alloc_0) : (memref<1024xf32>, memref<32x32xf32>) -> ()
    %alloc_1 = memref.alloc() : memref<32x32xf32>
    call @load_buf2(%arg2, %alloc_1) : (memref<1024xf32>, memref<32x32xf32>) -> ()
    call @gemm_0_0(%alloc, %alloc_0, %alloc_1) : (memref<32x32xf32>, memref<32x32xf32>, memref<32x32xf32>) -> ()
    call @gemm_0_1(%alloc, %alloc_0, %alloc_1) : (memref<32x32xf32>, memref<32x32xf32>, memref<32x32xf32>) -> ()
    call @gemm_1_0(%alloc, %alloc_0, %alloc_1) : (memref<32x32xf32>, memref<32x32xf32>, memref<32x32xf32>) -> ()
    call @gemm_1_1(%alloc, %alloc_0, %alloc_1) {last} : (memref<32x32xf32>, memref<32x32xf32>, memref<32x32xf32>) -> ()
    call @store_res(%alloc_1, %arg2) : (memref<32x32xf32>, memref<1024xf32>) -> ()
    return
  }
}
```

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [IR], [Builder], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage (It would be better to provide ~2 different test cases to test the robustness of your code)
- [x] Code is well-documented

### CODE REVIEW

**Comment by AdrianLiu00** [comment]
I'd like to first verify the functionality to see if any issues here. And then I'll remove the redundant code to streamline the passes and transforms.

File: allo/ir/transform.py [hunk 1]
```diff
@@ -299,40 +300,33 @@ def update_streaming_interface(module, target, depth=-1):
...
-        tensor.replace_all_uses_with(alloc_op.result)
+def create_data_movement(
+    tensors, name, ip, from_memory=True, flatten=False, mapping=None
+):
+
+    if len(tensors) != 2:
+        raise IndexError("One origin and one destination ONLY!")
+
+    tensor_ori = tensors[0]
+    tensor_des = tensors[1]
```
Comment (chhzh123) [inline allo/ir/transform.py]: `src_tensor`, `des_tensor`

File: allo/passes.py [hunk 2]
```diff
@@ -107,77 +108,195 @@ def annotate_affine_for(op):
...
+
+    # Modify Top function
+    with top_func.context, Location.unknown():
+        ip_first = InsertionPoint(top_func.entry_block.operations[0])
+
+        # Modify Loading
+        new_in_types = []
+        last_buf = None  # For Default Storing
+        with ip_first:
+            for ind_arg, arg in enumerate(top_func.arguments):
```
Comment (chhzh123) [inline allo/passes.py]: `arg_idx`

File: tests/dataflow/test_wrap_movement.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        module = str(mod.module)
+        assert (
+            f"call @load_buf0(%arg0, %alloc) : (memref<1024xf32>, memref<32x32xf32>) -> ()"
+            in module
+        )
+        assert (
+            f"call @store_res(%alloc_1, %arg2) : (memref<32x32xf32>, memref<1024xf32>) -> ()"
+            in module
+        )
+        print("Passed!")
```
Comment (chhzh123) [inline tests/dataflow/test_wrap_movement.py]: Move these assertions outside so that the CI is able to test them

Comment (chhzh123) [reply tests/dataflow/test_wrap_movement.py]: If `df.build(top)` cannot pass without vitis_hls environment, you can try the LLVM or `vhls` backend.

Comment (chhzh123) [reply tests/dataflow/test_wrap_movement.py]: Or is this feature only enabled with the vitis_hls backend?

**Review by chhzh123 — COMMENTED** [review]
Thanks for making this work! Mostly LGTM! Just some minor issues

**Comment by AdrianLiu00** [comment]
Data movement generation and wrapping are implemented exclusively for the _vitis_hls_ target in the actual FPGA flow. But I have exposed the pass in the test case to evaluate its functionality, enabling it to work in the CI pipeline independently of the Vitis environment. Besides, I have expanded test coverage to include Function Call, Buffer Allocation, and other related changes in this PR. Naming issues have also been resolved.

**Review by chhzh123 — APPROVED** [review]
Great! Thanks for resolving the issues!

**Pull request merged**